rubix-hadoop2/rubix-hadoop2CM/src/main/java/com/qubole/rubix/hadoop2/hadoop2CM/Hadoop2ClusterManager.java

```diff
@@ -46,23 +49,27 @@
     private boolean isMaster = true;
     public int serverPort = 8088;
     private String serverAddress = "localhost";
-    private Supplier<List<String>> nodesSupplier;
+    LoadingCache<String, List<String>> nodesCache;
+    YarnConfiguration yconf;
     String address = "localhost:8088";
     private Log log = LogFactory.getLog(Hadoop2ClusterManager.class);
     static String addressConf = "yarn.resourcemanager.webapp.address";
+    int refreshTime = super.getNodeRefreshTime();
 
     @Override
     public void initialize(Configuration conf)
     {
         super.initialize(conf);
-        log.debug("Initializing: Hadoop2ClusterManager");
-        this.address = conf.get(addressConf, address);
+        yconf = new YarnConfiguration();
+        log.info("Initializing: Hadoop2ClusterManager");
+        this.address = yconf.get(addressConf, address);
         this.serverAddress = address.substring(0, address.indexOf(":"));
         this.serverPort = Integer.parseInt(address.substring(address.indexOf(":") + 1));
-        nodesSupplier = Suppliers.memoizeWithExpiration(new Supplier<List<String>>()
+        nodesCache = CacheBuilder.newBuilder().refreshAfterWrite(refreshTime, TimeUnit.SECONDS).build(new CacheLoader<String, List<String>>()
```

Comment:
Need to do a async refresh: http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/cache/CacheLoader.html#reload(K,%20V)

http://www.theotherian.com/2013/11/non-blocking-cache-with-guava-and-listenable-futures.html